Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs

```diff
@@ -0,0 +1,182 @@
+                });
+
+                _exportDataSourceStates.Add(new ExportDataSourceState
+                {
+                    SearchCriteria = new PropertyDictionaryItemSearchCriteria() { CatalogIds = _dataQuery.CatalogIds },
+                    FetchFunc = async (x) =>
+                    {
+                        var searchResult = await _propertyDictionarySearchService.SearchAsync((PropertyDictionaryItemSearchCriteria)x.SearchCriteria);
+                        x.TotalCount = searchResult.TotalCount;
+                        x.Result = searchResult.Results;
+                    }
+                });
+
+                _exportDataSourceStates.Add(new ExportDataSourceState
+                {
+                    SearchCriteria = new ProductSearchCriteria() { CatalogIds = _dataQuery.CatalogIds, ResponseGroup = (ItemResponseGroup.Full & ~ItemResponseGroup.Variations).ToString() },
+                    FetchFunc = async (x) =>
+                    {
+                        var searchResult = await _productSearchService.SearchProductsAsync((ProductSearchCriteria)x.SearchCriteria);
+                        x.TotalCount = searchResult.TotalCount;
+                        x.Result = searchResult.Results;
+                    }
+                });
+
+                CalculateCounts();
+            }
+        }
+
+        public int GetTotalCount()
+        {
+            CalculateCounts();
+            return TotalCount;
+        }
+
+        private void EnsureHaveTotals()
+        {
+            if (TotalCount < 0)
+            {
+                CalculateCounts();
+            }
+        }
+
+        public bool Fetch()
+        {
+            EnsureHaveTotals();
+
+            var take = _dataQuery.Take ?? PageSize;
+            var skip = _dataQuery.Skip ?? CurrentPageNumber * PageSize;
+            var taskList = new List<Task>();
+            Items = new List<IExportable>();
+
+            var curStateStartPos = 0; // Starting position of current state
+
+            foreach (var state in _exportDataSourceStates)
+            {
+                state.Result = Array.Empty<IExportable>();
+                var overlap = skip - curStateStartPos; // Positive overlap should fall into skip, negative into take
+                state.SearchCriteria.Skip = overlap < 0 ? 0 : overlap;
+                state.SearchCriteria.Take = (skip + take > curStateStartPos + state.TotalCount) ? curStateStartPos + state.TotalCount - skip : take + (overlap < 0 ? overlap : 0);
+
+                if (state.SearchCriteria.Take > 0) // Skip running this fetch in case of nothing to take
+                {
+                    taskList.Add(state.FetchFunc(state));
+                }
+
+                curStateStartPos += state.TotalCount;
+            }
+
+            Task.WhenAll(taskList).GetAwaiter().GetResult(); // Ensures completing all fetches
+
+            ((List<IExportable>)Items).AddRange(_exportDataSourceStates.SelectMany(x => x.Result));
```

Comment:
Why do we need this cast?

Reply:
Ok. I'll transform it to Items.Concat...

Reply:
Ok. Removed.